Mathlib/Algebra/Homology/EulerCharacteristic.lean

```diff
@@ -0,0 +1,150 @@
+
+/-- Signs alternate in the backward direction of the complex shape. -/
+lemma χ_prev {i j : ι} (h : c.Rel i j) : c.χ i = - c.χ j := by simp [c.χ_next h]
+
+@[simps]
+instance eulerCharSignsUpInt : (up ℤ).EulerCharSigns where
+  χ := Int.negOnePow
+  χ_next := by rintro _ _ rfl; rw [Int.negOnePow_succ]
+
+@[simps]
+instance eulerCharSignsDownInt : (down ℤ).EulerCharSigns where
+  χ := Int.negOnePow
+  χ_next := by rintro _ _ rfl; simp [Int.negOnePow_succ]
+
+@[simps]
+instance eulerCharSignsUpNat : (up ℕ).EulerCharSigns where
+  χ n := (-1) ^ n
+  χ_next := by rintro _ _ rfl; simp [pow_add]
+
+@[simps]
+instance eulerCharSignsDownNat : (down ℕ).EulerCharSigns where
+  χ n := (-1) ^ n
+  χ_next := by rintro _ _ rfl; simp [pow_add]
+
+end ComplexShape
+
+open ComplexShape
+
+variable {R : Type*} [Ring R] {ι : Type*} {c : ComplexShape ι}
+
+namespace HomologicalComplex
+
+section FiniteSum
+
+/-- The Euler characteristic of a homological complex over a finite set of indices.
+The sign at each index is determined by the `ComplexShape.EulerCharSigns` instance. -/
+noncomputable def eulerChar [c.EulerCharSigns] (C : HomologicalComplex (ModuleCat R) c)
+    (indices : Finset ι) : ℤ :=
+  ∑ i ∈ indices, (c.χ i : ℤ) * Module.finrank R (C.X i)
+
+/-- The homological Euler characteristic over a finite set of indices.
+This is the Euler characteristic computed from the homology groups rather than
+the original complex. -/
+noncomputable def homologyEulerChar [c.EulerCharSigns]
+    (C : HomologicalComplex (ModuleCat R) c) [∀ i : ι, C.HasHomology i]
+    (indices : Finset ι) : ℤ :=
+  ∑ i ∈ indices, (c.χ i : ℤ) * Module.finrank R (C.homology i)
+
+end FiniteSum
+
+section InfiniteSum
+
+variable [Fintype ι]
+
+/-- The Euler characteristic as an infinite sum over all indices.
+This requires the index type to be finite. -/
+noncomputable def eulerCharTsum [c.EulerCharSigns]
+    (C : HomologicalComplex (ModuleCat R) c) : ℤ :=
+  ∑ i : ι, (c.χ i : ℤ) * Module.finrank R (C.X i)
+
+/-- The homological Euler characteristic as an infinite sum. -/
+noncomputable def homologyEulerCharTsum [c.EulerCharSigns]
+    (C : HomologicalComplex (ModuleCat R) c) [∀ i : ι, C.HasHomology i] : ℤ :=
+  ∑ i : ι, (c.χ i : ℤ) * Module.finrank R (C.homology i)
+
+/-- If a complex vanishes outside a finite set, the infinite Euler characteristic
+equals the finite one. -/
+theorem eulerCharTsum_eq_eulerChar [c.EulerCharSigns]
+    (C : HomologicalComplex (ModuleCat R) c)
+    (indices : Finset ι)
+    (h_zero : ∀ i ∉ indices, Module.finrank R (C.X i) = 0) :
+    eulerCharTsum C = eulerChar C indices := by
+  simp only [eulerCharTsum, eulerChar]
+  symm
+  apply Finset.sum_subset (Finset.subset_univ indices)
+  intro i _ hi
+  simp [h_zero i hi]
+
+/-- If homology vanishes outside a finite set, the infinite homological Euler
+characteristic equals the finite one. -/
+theorem homologyEulerCharTsum_eq_homologyEulerChar [c.EulerCharSigns]
```

Comment:
Could you introduce definitions similar to `eulerChar` and `eulerCharTsum` but for `GradedObject`? (This would require adding `c` as an explicit parameter.)
Then, you may introduce abbrevs for these two definitions in the particular case of `C.X` and `C.homology`.
The lemma `homologyEulerCharTsum_eq_homologyEulerChar` should be stated for this more general definition on graded objects.

Reply:
Great idea! Done.